crypto/src/main/scala/org/bitcoins/crypto/BouncyCastleUtil.scala

```diff
@@ -215,3 +220,224 @@ object BouncyCastleUtil {
+    val adaptedSig = FieldElement(adaptorSig.adaptedSig.drop(33))
+
+    val validProof = DLEQStuff.dleqVerify(
+      "ECDSAAdaptorSig",
+      proofS,
+      proofR,
+      untweakedNonce,
+      adaptor,
+      tweakedNonce
+    )
+
+    if (validProof) {
+      val tweakedNoncex = FieldElement(tweakedNonce.bytes.tail)
+      val untweakedNoncex = FieldElement(untweakedNonce.bytes.tail)
+
+      if (tweakedNoncex.isZero || untweakedNoncex.isZero) {
+        false
+      } else {
+
+        val untweakedRx =
+          adaptorVerifyHelper(tweakedNoncex, adaptedSig, pubKey, data)
+
+        untweakedRx == untweakedNoncex
+      }
+    } else {
+      false
+    }
+  }
+
+  def adaptorComplete(
+      adaptorSecret: ECPrivateKey,
+      adaptedSig: ByteVector): ECDigitalSignature = {
+    val tweakedNonce: ECPublicKey =
+      ECAdaptorSignature.deserializePoint(adaptedSig.take(33))
+    val rx = FieldElement(tweakedNonce.bytes.tail)
+    val adaptedS: FieldElement = FieldElement(adaptedSig.drop(33))
+    val correctedS = adaptedS.multInv(adaptorSecret.fieldElement)
+
+    val sig = ECDigitalSignature.fromRS(BigInt(rx.toBigInteger),
+                                        BigInt(correctedS.toBigInteger))
+    DERSignatureUtil.lowS(sig)
+  }
+
+  def extractAdaptorSecret(
+      sig: ECDigitalSignature,
+      adaptorSig: ECAdaptorSignature,
+      adaptor: ECPublicKey): ECPrivateKey = {
+    val secretOrNeg = adaptorSig.adaptedS.multInv(FieldElement(sig.s))
+    if (secretOrNeg.getPublicKey == adaptor) {
+      secretOrNeg.toPrivateKey
+    } else {
+      secretOrNeg.negate.toPrivateKey
+    }
+  }
+}
+
+object DLEQStuff {
+  import ECAdaptorSignature.serializePoint
+
+  def dleqPair(
+      fe: FieldElement,
+      adaptorPoint: ECPublicKey): (ECPublicKey, ECPublicKey) = {
+    val point = fe.getPublicKey
+    val tweakedPoint = adaptorPoint.tweakMultiply(fe)
+
+    (point, tweakedPoint)
+  }
+
+  def dleqNonceFunc(
+      hash: ByteVector,
+      fe: FieldElement,
+      algoName: String): FieldElement = {
+    val kBytes =
+      CryptoUtil.taggedSha256(fe.bytes ++ hash, algoName).bytes
+    FieldElement(kBytes)
+  }
+
+  def dleqChallengeHash(
+      algoName: String,
+      adaptorPoint: ECPublicKey,
+      r1: ECPublicKey,
+      r2: ECPublicKey,
+      p1: ECPublicKey,
+      p2: ECPublicKey): ByteVector = {
+    CryptoUtil
+      .taggedSha256(
+        serializePoint(adaptorPoint) ++ serializePoint(r1) ++ serializePoint(
+          r2) ++ serializePoint(p1) ++ serializePoint(p2),
+        algoName)
+      .bytes
+  }
+
+  /** Proves that the DLOG_G(R') = DLOG_Y(R) (= fe)
+    * For a full description, see https://cs.nyu.edu/courses/spring07/G22.3220-001/lec3.pdf
+    */
+  def dleqProve(
+      fe: FieldElement,
+      adaptorPoint: ECPublicKey,
+      algoName: String): (FieldElement, FieldElement) = {
+    // (fe*G, fe*Y)
+    val (p1, p2) = dleqPair(fe, adaptorPoint)
```

Comment:
nit: name these instead `feG, feY`

Reply:
I believe these will be the names used in the spec (as it is what the C impl does) but I could be wrong